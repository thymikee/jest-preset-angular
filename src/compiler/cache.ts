import type * as ts from 'typescript';

/**
 * Copy from https://github.com/angular/angular-cli/blob/master/packages/ngtools/webpack/src/ivy/cache.ts
 */
export class SourceFileCache extends Map<string, ts.SourceFile> {
  readonly #angularDiagnostics = new Map<ts.SourceFile, ts.Diagnostic[]>();

  invalidate(file: string): void {
    const sourceFile = this.get(file);
    if (sourceFile) {
      this.delete(file);
      this.#angularDiagnostics.delete(sourceFile);
    }
  }

  updateAngularDiagnostics(sourceFile: ts.SourceFile, diagnostics: ts.Diagnostic[]): void {
    if (diagnostics.length > 0) {
      this.#angularDiagnostics.set(sourceFile, diagnostics);
    } else {
      this.#angularDiagnostics.delete(sourceFile);
    }
  }

  getAngularDiagnostics(sourceFile: ts.SourceFile): ts.Diagnostic[] | undefined {
    return this.#angularDiagnostics.get(sourceFile);
  }
}
