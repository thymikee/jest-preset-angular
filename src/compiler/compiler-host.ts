import type { Logger } from 'bs-logger';
import { normalize } from 'path';
import type { TTypeScript } from 'ts-jest/dist/types';
import { LINE_FEED } from 'ts-jest/dist/constants';
import type * as ts from 'typescript';

import type { NgJestConfig } from '../config/ng-jest-config';
import type { CompilerOptions } from '@angular/compiler-cli';

export class NgJestCompilerHost implements ts.CompilerHost {
  private readonly _sourceFileCache: Map<string, ts.SourceFile> = new Map<string, ts.SourceFile>();
  private readonly _memoryHost: Map<string, string | undefined> = new Map<string, string | undefined>();
  private readonly _emittedResult: [string, string] = ['', ''];
  private readonly _ts: TTypeScript;
  private readonly _moduleResolutionCache: ts.ModuleResolutionCache;

  constructor(private logger: Logger, private ngJestCfg: NgJestConfig) {
    this._ts = this.ngJestCfg.compilerModule;
    this._moduleResolutionCache = this._ts.createModuleResolutionCache(this.ngJestCfg.cwd, (x) => x);
  }

  getEmittedResult(): [string, string] {
    return this._emittedResult;
  }

  updateMemoryHost(fileName: string, fileContent: string): void {
    const previousContents = this._memoryHost.get(fileName);
    const contentsChanged = previousContents !== fileContent;
    if (contentsChanged) {
      this._memoryHost.set(fileName, fileContent);
    }
  }

  getSourceFile(fileName: string, languageVersion: ts.ScriptTarget): ts.SourceFile | undefined {
    const normalizedFileName = normalize(fileName);
    try {
      const cached = this._sourceFileCache.get(normalizedFileName);
      if (cached) {
        return cached;
      }

      this.logger.debug(
        { normalizedFileName },
        'getSourceFile: cache miss, reading file and create source file to update cache',
      );

      const content = this.readFile(normalizedFileName);
      if (content !== undefined) {
        const sf = this._ts.createSourceFile(normalizedFileName, content, languageVersion, true);
        this._sourceFileCache.set(fileName, sf);

        return sf;
      }
    } catch (e) {
      this.logger.error(e);
    }

    return undefined;
  }

  getDefaultLibFileName(options: ts.CompilerOptions): string {
    return this._ts.createCompilerHost(options).getDefaultLibFileName(options);
  }

  // This is due to typescript CompilerHost interface being weird on writeFile. This shuts down
  // typings in WebStorm.
  get writeFile() {
    return (fileName: string, data: string, _writeByteOrderMark: boolean): void => {
      this._emittedResult[fileName.endsWith('.map') ? 1 : 0] = data;
    };
  }

  getCurrentDirectory(): string {
    return this.ngJestCfg.cwd;
  }

  getCanonicalFileName(fileName: string): string {
    return this.useCaseSensitiveFileNames() ? fileName : fileName.toLowerCase();
  }

  useCaseSensitiveFileNames(): boolean {
    return this._ts.sys.useCaseSensitiveFileNames;
  }

  getNewLine(): string {
    return LINE_FEED;
  }

  fileExists(fileName: string): boolean {
    return this._ts.sys.fileExists(fileName);
  }

  readFile(fileName: string): string | undefined {
    const normalizedFileName = normalize(fileName);
    let fileContent = this._memoryHost.get(normalizedFileName);
    if (!fileContent) {
      this.logger.debug(
        { fileName: normalizedFileName },
        'readFile: file does not exist in memory cache, read file with file system',
      );

      fileContent = this._ts.sys.readFile(normalizedFileName) ?? undefined;
      this._memoryHost.set(normalizedFileName, fileContent);
    }

    return fileContent;
  }

  trace(message: string): void {
    this.logger.trace(message);
  }

  resolveModuleNames(
    moduleNames: string[],
    containingFile: string,
    _reusedNames: string[] | undefined,
    redirectedReference: ts.ResolvedProjectReference | undefined,
    options: CompilerOptions,
  ): (ts.ResolvedModule | undefined)[] {
    return moduleNames.map((moduleName) => {
      const { resolvedModule } = this._ts.resolveModuleName(
        moduleName,
        containingFile,
        options,
        this._ts.createCompilerHost(options),
        this._moduleResolutionCache,
        redirectedReference,
      );

      return resolvedModule;
    });
  }

  resolveTypeReferenceDirectives(
    typeReferenceDirectiveNames: string[],
    containingFile: string,
    redirectedReference: ts.ResolvedProjectReference | undefined,
    options: CompilerOptions,
  ): (ts.ResolvedTypeReferenceDirective | undefined)[] {
    return typeReferenceDirectiveNames.map((typeDirectiveName) => {
      const { resolvedTypeReferenceDirective } = this._ts.resolveTypeReferenceDirective(
        typeDirectiveName,
        containingFile,
        options,
        this._ts.createCompilerHost(options),
        redirectedReference,
      );

      return resolvedTypeReferenceDirective;
    });
  }
}
