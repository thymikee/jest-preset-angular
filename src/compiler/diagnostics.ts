import type { Diagnostic, Diagnostics } from '@angular/compiler-cli';
import { CancellationToken, DiagnosticMode, hasErrors } from '@ngtools/webpack/src/diagnostics';
import type * as ts from 'typescript';

/**
 * Copy from `@ngtools/webpack` and strip out stuffs
 */
export function gatherDiagnostics(
  program: ts.Program,
  mode = DiagnosticMode.All,
  cancellationToken?: CancellationToken,
): Diagnostics {
  const allDiagnostics: Array<ts.Diagnostic | Diagnostic> = [];
  let checkOtherDiagnostics = true;

  // eslint-disable-next-line @typescript-eslint/ban-types
  function checkDiagnostics<T extends Function>(fn: T) {
    if (checkOtherDiagnostics) {
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
      const diags = fn(undefined, cancellationToken);
      if (diags) {
        allDiagnostics.push(...diags);

        checkOtherDiagnostics = !hasErrors(diags);
      }
    }
  }

  const gatherSyntacticDiagnostics = (mode & DiagnosticMode.Syntactic) != 0;
  const gatherSemanticDiagnostics = (mode & DiagnosticMode.Semantic) != 0;
  const tsProgram = program;
  if (gatherSyntacticDiagnostics) {
    // Check syntactic diagnostics.
    checkDiagnostics(tsProgram.getSyntacticDiagnostics.bind(tsProgram));
  }
  if (gatherSemanticDiagnostics) {
    checkDiagnostics(tsProgram.getSemanticDiagnostics.bind(tsProgram));
  }

  return allDiagnostics;
}
