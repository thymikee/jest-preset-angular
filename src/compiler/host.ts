import ts from 'typescript';

import { normalizePath } from './paths';

export const augmentHostWithCaching = (host: ts.CompilerHost, cache: Map<string, ts.SourceFile>): void => {
  const baseGetSourceFile = host.getSourceFile;
  host.getSourceFile = function (fileName, languageVersion, onError, shouldCreateNewSourceFile, ...parameters) {
    if (!shouldCreateNewSourceFile && cache.has(fileName)) {
      return cache.get(fileName);
    }

    const file = baseGetSourceFile.call(host, fileName, languageVersion, onError, true, ...parameters);

    if (file) {
      cache.set(fileName, file);
    }

    return file;
  };
};

/**
 * Augments a TypeScript Compiler Host's resolveModuleNames function to collect dependencies
 * of the containing file passed to the resolveModuleNames function. This process assumes
 * that consumers of the Compiler Host will only call resolveModuleNames with modules that are
 * actually present in a containing file.
 * This process is a workaround for gathering a TypeScript SourceFile's dependencies as there
 * is no currently exposed public method to do so. A BuilderProgram does have a `getAllDependencies`
 * function.
 */
export const augmentHostWithDependencyCollection = (
  host: ts.CompilerHost,
  dependencies: Map<string, Set<string>>,
  moduleResolutionCache?: ts.ModuleResolutionCache,
): void => {
  if (host.resolveModuleNames) {
    const baseResolveModuleNames = host.resolveModuleNames;
    host.resolveModuleNames = function (moduleNames: string[], containingFile: string, ...parameters) {
      const results = baseResolveModuleNames.call(host, moduleNames, containingFile, ...parameters);

      const containingFilePath = normalizePath(containingFile);
      for (const result of results) {
        if (result) {
          const containingFileDependencies = dependencies.get(containingFilePath);
          if (containingFileDependencies) {
            containingFileDependencies.add(result.resolvedFileName);
          } else {
            dependencies.set(containingFilePath, new Set([result.resolvedFileName]));
          }
        }
      }

      return results;
    };
  } else {
    host.resolveModuleNames = function (
      moduleNames: string[],
      containingFile: string,
      _reusedNames: string[] | undefined,
      redirectedReference: ts.ResolvedProjectReference | undefined,
      options: ts.CompilerOptions,
    ) {
      return moduleNames.map((name) => {
        const result = ts.resolveModuleName(
          name,
          containingFile,
          options,
          host,
          moduleResolutionCache,
          redirectedReference,
        ).resolvedModule;

        if (result) {
          const containingFilePath = normalizePath(containingFile);
          const containingFileDependencies = dependencies.get(containingFilePath);
          if (containingFileDependencies) {
            containingFileDependencies.add(result.resolvedFileName);
          } else {
            dependencies.set(containingFilePath, new Set([result.resolvedFileName]));
          }
        }

        return result;
      });
    };
  }
};
