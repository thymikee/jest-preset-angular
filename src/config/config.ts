import type { TsJestTransformerOptions } from 'ts-jest';

export type NgJestTransformerOptions = TsJestTransformerOptions & {
    processWithEsbuild?: string[];
};
