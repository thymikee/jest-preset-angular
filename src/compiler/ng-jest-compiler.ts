import { CompilerHost, CompilerOptions, createCompilerHost } from '@angular/compiler-cli';
import { hasErrors } from '@ngtools/webpack/src/diagnostics';
import type { Logger } from 'bs-logger';
import { updateOutput } from 'ts-jest/dist/compiler/compiler-utils';
import type { CompilerInstance, TTypeScript, ResolvedModulesMap } from 'ts-jest/dist/types';
import type * as ts from 'typescript';

import type { NgJestConfig } from '../config/ng-jest-config';
import { factory as downlevelCtor } from '../transformers/downlevel-ctor';
import { NgJestCompilerHost } from './compiler-host';

export class NgJestCompiler implements CompilerInstance {
  private _compilerOptions!: CompilerOptions;
  private _program: ts.Program | undefined;
  private _compilerHost: CompilerHost | undefined;
  private _tsHost: NgJestCompilerHost | undefined;
  private _rootNames: string[] = [];
  private readonly _logger: Logger;
  private readonly _ts: TTypeScript;

  constructor(readonly ngJestConfig: NgJestConfig) {
    this._logger = this.ngJestConfig.logger;
    this._ts = this.ngJestConfig.compilerModule;
    this._setupOptions(this.ngJestConfig);

    this._logger.debug('created NgJestCompiler');
  }

  getResolvedModulesMap(fileContent: string, fileName: string): ResolvedModulesMap {
    this._tsHost?.updateMemoryHost(fileName, fileContent);

    // eslint-disable-next-line
    return (this._program?.getSourceFile(fileName) as any)?.resolvedModules;
  }

  getCompiledOutput(fileName: string, fileContent: string): string {
    if (this._program) {
      const allDiagnostics = [];
      if (!this._rootNames.includes(fileName)) {
        this._logger.debug({ fileName }, 'getCompiledOutput: update memory host, rootFiles and Program');

        this._rootNames = [...this._rootNames, fileName];
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        this._tsHost!.updateMemoryHost(fileName, fileContent);
        this._createOrUpdateProgram();
      }

      this._logger.debug({ fileName }, 'getCompiledOutput: compiling using Program');

      const sourceFile = this._program.getSourceFile(fileName);
      const emitResult = this._program.emit(sourceFile, undefined, undefined, undefined, {
        before: [
          // hoisting from `ts-jest` or other before transformers
          ...(this.ngJestConfig.customTransformers.before as ts.TransformerFactory<ts.SourceFile>[]),
          /**
           * Downlevel constructor parameters for DI support. This is required to support forwardRef in ES2015 due to
           * TDZ issues. This wrapper is needed here due to the program not being available until after
           * the transformers are created. Also because program can be updated so we can't push this transformer in
           * _createCompilerHost
           */
          downlevelCtor(this._program),
        ],
      });
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      const compiledOutput: [string, string] = this._tsHost!.getEmittedResult();
      if (this.ngJestConfig.shouldReportDiagnostics(fileName)) {
        this._logger.debug({ fileName }, 'getCompiledOutput: getting diagnostics');

        allDiagnostics.push(
          ...emitResult.diagnostics,
          ...this._program.getSyntacticDiagnostics(sourceFile),
          ...this._program.getSemanticDiagnostics(sourceFile),
        );
      }
      if (!hasErrors(allDiagnostics)) {
        this._logger.debug({ fileName }, 'getCompiledOutput: update compiled output including inline source map');

        return updateOutput(compiledOutput[0], fileName, compiledOutput[1]);
      } else {
        this.ngJestConfig.raiseDiagnostics(allDiagnostics, fileName);

        return '';
      }
    } else {
      this._logger.debug({ fileName }, 'getCompiledOutput: compiling as isolated module');

      const result: ts.TranspileOutput = this._ts.transpileModule(fileContent, {
        fileName,
        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
        transformers: this.ngJestConfig.customTransformers,
        compilerOptions: this._compilerOptions,
        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
        reportDiagnostics: this.ngJestConfig.shouldReportDiagnostics(fileName),
      });
      if (result.diagnostics && this.ngJestConfig.shouldReportDiagnostics(fileName)) {
        this.ngJestConfig.raiseDiagnostics(result.diagnostics, fileName, this._logger);
      }

      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      return updateOutput(result.outputText, fileName, result.sourceMapText!);
    }
  }

  private _setupOptions({ parsedTsConfig }: NgJestConfig) {
    this._logger.debug({ parsedTsConfig }, '_setupOptions: initializing compiler config');

    this._compilerOptions = { ...parsedTsConfig.options };
    this._rootNames = parsedTsConfig.rootNames.filter((rootName) => !this.ngJestConfig.isTestFile(rootName));
    if (this._compilerOptions.strictMetadataEmit) {
      this._logger.warn(
        `Using Angular compiler option 'strictMetadataEmit' for applications might cause undefined behavior.`,
      );
    }
    if (!this.ngJestConfig.isolatedModules) {
      this._logger.debug(
        { compilerOptions: this._compilerOptions },
        '_setupOptions: creating Compiler Host using @angular/compiler-cli createCompilerHost',
      );

      this._tsHost = new NgJestCompilerHost(this._logger, this.ngJestConfig);
      this._compilerHost = createCompilerHost({
        options: this._compilerOptions,
        tsHost: this._tsHost,
      });
      this._createOrUpdateProgram();
    }
  }

  private _createOrUpdateProgram(): void {
    const oldTsProgram = this._program;

    this._logger.debug(`_createOrUpdateProgram: ${oldTsProgram ? 'update' : 'create'} TypeScript Program`);

    this._program = this._ts.createProgram(this._rootNames, this._compilerOptions, this._compilerHost, oldTsProgram);
  }
}
