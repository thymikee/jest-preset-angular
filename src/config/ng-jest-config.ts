import type { Logger } from 'bs-logger';
import { globsToMatcher } from 'jest-util';
import type { ProjectConfigTsJest, RawCompilerOptions } from 'ts-jest';
import { ConfigSet } from 'ts-jest/dist/legacy/config/config-set';
import type { ParsedCommandLine } from 'typescript';

/**
 * Some NPM packages like `tslib` is distributed in such a way that `esbuild` cannot process it, so we fall back to use TypeScript API
 */
const defaultProcessWithEsbuildPatterns = ['**/*.mjs'];

export class NgJestConfig extends ConfigSet {
  readonly processWithEsbuild: ReturnType<typeof globsToMatcher>;

  constructor(jestConfig: ProjectConfigTsJest | undefined, parentLogger?: Logger | undefined) {
    super(jestConfig, parentLogger);
    const jestGlobalsConfig = jestConfig?.globals?.ngJest ?? Object.create(null);
    this.processWithEsbuild = globsToMatcher([
      ...(jestGlobalsConfig.processWithEsbuild ?? []),
      ...defaultProcessWithEsbuildPatterns,
    ]);
  }

  /**
   * Override `ts-jest` behavior because we use `readConfiguration` which will read and resolve tsconfig.
   */
  protected _resolveTsConfig(compilerOptions?: RawCompilerOptions, resolvedConfigFile?: string): ParsedCommandLine {
    const result = super._resolveTsConfig(compilerOptions, resolvedConfigFile) as ParsedCommandLine;
    result.options.enableIvy = true;
    result.options.noEmitOnError = false;
    result.options.suppressOutputPathCheck = true;
    result.options.allowEmptyCodegenFiles = false;
    result.options.annotationsAs = 'decorators';
    result.options.enableResourceInlining = false;
    // Since we define preset default also transform `js` so we need to set `allowJs` true
    result.options.allowJs = true;
    const ts = this.compilerModule;
    const scriptTarget = result.options.target ?? ts.ScriptTarget?.ES2015;
    if (scriptTarget > ts.ScriptTarget?.ES2016) {
      // See https://github.com/angular/components/issues/21632#issuecomment-764975917
      result.options.target = ts.ScriptTarget?.ES2015;
    }

    return result;
  }
}
