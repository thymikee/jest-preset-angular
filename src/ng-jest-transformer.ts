import path from 'path';

import type { TransformedSource } from '@jest/transform';
import type { Config } from '@jest/types';
import { LogContexts, LogLevels, Logger, createLogger } from 'bs-logger';
import { transformSync } from 'esbuild';
import { ConfigSet } from 'ts-jest/dist/config/config-set';
import { TsJestTransformer } from 'ts-jest/dist/ts-jest-transformer';
import type { ProjectConfigTsJest, TransformOptionsTsJest } from 'ts-jest/dist/types';

import { NgJestCompiler } from './compiler/ng-jest-compiler';
import { NgJestConfig } from './config/ng-jest-config';

export class NgJestTransformer extends TsJestTransformer {
  #ngJestLogger: Logger;

  constructor() {
    super();
    this.#ngJestLogger = createLogger({
      context: {
        [LogContexts.package]: 'ts-jest',
        [LogContexts.logLevel]: LogLevels.trace,
        // eslint-disable-next-line @typescript-eslint/no-var-requires
        version: require('../package.json').version,
      },
      targets: process.env.NG_JEST_LOG ?? undefined,
    });
  }

  protected _createConfigSet(config: ProjectConfigTsJest | undefined): ConfigSet {
    return new NgJestConfig(config);
  }

  protected _createCompiler(configSet: ConfigSet, cacheFS: Map<string, string>): void {
    this._compiler = new NgJestCompiler(configSet, cacheFS);
  }

  process(
    fileContent: string,
    filePath: Config.Path,
    transformOptions: TransformOptionsTsJest,
  ): TransformedSource | string {
    const configSet = this._createConfigSet(transformOptions.config);
    /**
     * TypeScript < 4.5 doesn't support compiling `.mjs` file by default when running `tsc` which throws error. Also we
     * transform `js` files from `node_modules` assuming that `node_modules` contains compiled files to speed up compilation.
     * IMPORTANT: we exclude `tslib` from compilation because it has issue with compilation. The original `tslib.js` or
     * `tslib.es6.js` works well with Jest without extra compilation
     */
    if (
      path.extname(filePath) === '.mjs' ||
      (/node_modules\/(.*.js$)/.test(filePath.replace(/\\/g, '/')) && !filePath.includes('tslib'))
    ) {
      this.#ngJestLogger.debug({ filePath }, 'process with esbuild');

      const compilerOpts = configSet.parsedTsConfig.options;
      const { code, map } = transformSync(fileContent, {
        loader: 'js',
        format: transformOptions.supportsStaticESM && configSet.useESM ? 'esm' : 'cjs',
        target: compilerOpts.target === configSet.compilerModule.ScriptTarget.ES2015 ? 'es2015' : 'es2016',
        sourcemap: compilerOpts.sourceMap,
        sourcefile: filePath,
        sourcesContent: true,
        sourceRoot: compilerOpts.sourceRoot,
      });

      return {
        code,
        map,
      };
    } else {
      return super.process(fileContent, filePath, transformOptions);
    }
  }
}
