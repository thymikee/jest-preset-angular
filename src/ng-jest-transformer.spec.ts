import { transformSync } from 'esbuild';

import { NgJestCompiler } from './compiler/ng-jest-compiler';
import { NgJestConfig } from './config/ng-jest-config';
import { NgJestTransformer } from './ng-jest-transformer';

const tr = new NgJestTransformer();

jest.mock('esbuild', () => {
  return {
    transformSync: jest.fn().mockReturnValue({
      code: '',
      map: '',
    }),
  };
});

describe('NgJestTransformer', () => {
  test('should create NgJestCompiler and NgJestConfig instances', () => {
    // @ts-expect-error testing purpose
    const cs = tr._createConfigSet({
      cwd: process.cwd(),
      extensionsToTreatAsEsm: [],
      testMatch: [],
      testRegex: [],
    });

    // @ts-expect-error testing purpose
    tr._createCompiler(cs, new Map());

    // @ts-expect-error testing purpose
    expect(tr._compiler).toBeInstanceOf(NgJestCompiler);
    expect(cs).toBeInstanceOf(NgJestConfig);
  });

  test('should not use esbuild to process js files which are not from `node_modules`', () => {
    tr.process(
      `
      const pi = parseFloat(3.124);
      
      export { pi };
    `,
      'foo.js',
      {
        config: {
          cwd: process.cwd(),
          extensionsToTreatAsEsm: [],
          testMatch: [],
          testRegex: [],
          globals: {
            'ts-jest': {
              isolatedModules: true,
            },
          },
        },
      } as any, // eslint-disable-line @typescript-eslint/no-explicit-any,
    );

    // eslint-disable-next-line @typescript-eslint/no-explicit-any,
    expect(transformSync as unknown as jest.MockInstance<unknown, any>).not.toHaveBeenCalled();
  });

  test('should not use esbuild to process tslib file', () => {
    tr.process(
      `
      const pi = parseFloat(3.124);
      
      export { pi };
    `,
      'node_modules/tslib.es6.js',
      {
        config: {
          cwd: process.cwd(),
          extensionsToTreatAsEsm: [],
          testMatch: [],
          testRegex: [],
          globals: {
            'ts-jest': {
              isolatedModules: true,
            },
          },
        },
      } as any, // eslint-disable-line @typescript-eslint/no-explicit-any,
    );

    // eslint-disable-next-line @typescript-eslint/no-explicit-any,
    expect(transformSync as unknown as jest.MockInstance<unknown, any>).not.toHaveBeenCalled();
  });

  test.each([
    {
      tsconfig: {
        sourceMap: false,
      },
    },
    {
      tsconfig: {
        target: 'es2016',
      },
    },
    {
      tsconfig: {},
    },
  ])('should use esbuild to process mjs or `node_modules` js files to CJS codes', ({ tsconfig }) => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const transformSyncMock = transformSync as unknown as jest.MockInstance<unknown, any>;
    const transformCfg = {
      cacheFS: new Map(),
      config: {
        cwd: process.cwd(),
        extensionsToTreatAsEsm: [],
        testMatch: [],
        testRegex: [],
        globals: {
          'ts-jest': {
            tsconfig,
          },
          ngJest: {
            processWithEsbuild: ['node_modules/foo.js'],
          },
        },
      },
    } as any; // eslint-disable-line @typescript-eslint/no-explicit-any
    tr.process(
      `
      const pi = parseFloat(3.124);
      
      export { pi };
    `,
      'foo.mjs',
      transformCfg,
    );
    tr.process(
      `
      const pi = parseFloat(3.124);
      
      export { pi };
    `,
      'node_modules/foo.js',
      transformCfg,
    );

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    expect(transformSyncMock.mock.calls[0]).toMatchSnapshot();
    expect(transformSyncMock.mock.calls[1]).toMatchSnapshot();

    transformSyncMock.mockClear();
  });

  test.each([
    {
      tsconfig: {
        sourceMap: false,
      },
    },
    {
      tsconfig: {
        target: 'es2016',
      },
    },
    {
      tsconfig: {},
    },
  ])('should use esbuild to process mjs or `node_modules` js files to ESM codes', ({ tsconfig }) => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const transformSyncMock = transformSync as unknown as jest.MockInstance<unknown, any>;
    const transformCfg = {
      cacheFS: new Map(),
      config: {
        cwd: process.cwd(),
        extensionsToTreatAsEsm: [],
        testMatch: [],
        testRegex: [],
        globals: {
          'ts-jest': {
            tsconfig,
            useESM: true,
          },
          ngJest: {
            processWithEsbuild: ['node_modules/foo.js'],
          },
        },
      },
      supportsStaticESM: true,
    } as any; // eslint-disable-line @typescript-eslint/no-explicit-any
    tr.process(
      `
      const pi = parseFloat(3.124);
      
      export { pi };
    `,
      'foo.mjs',
      transformCfg,
    );
    tr.process(
      `
      const pi = parseFloat(3.124);
      
      export { pi };
    `,
      'node_modules/foo.js',
      transformCfg,
    );

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    expect(transformSyncMock.mock.calls[0]).toMatchSnapshot();
    expect(transformSyncMock.mock.calls[1]).toMatchSnapshot();

    transformSyncMock.mockClear();
  });
});
