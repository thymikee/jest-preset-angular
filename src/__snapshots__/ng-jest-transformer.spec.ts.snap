// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NgJestTransformer should use esbuild to process mjs or \`node_modules\` js files to CJS codes 1`] = `
Array [
  "
      const pi = parseFloat(3.124);
      
      export { pi };
    ",
  Object {
    "format": "cjs",
    "loader": "js",
    "sourceRoot": undefined,
    "sourcefile": "foo.mjs",
    "sourcemap": false,
    "sourcesContent": true,
    "target": "es2015",
  },
]
`;

exports[`NgJestTransformer should use esbuild to process mjs or \`node_modules\` js files to CJS codes 2`] = `
Array [
  "
      const pi = parseFloat(3.124);
      
      export { pi };
    ",
  Object {
    "format": "cjs",
    "loader": "js",
    "sourceRoot": undefined,
    "sourcefile": "node_modules/foo.js",
    "sourcemap": false,
    "sourcesContent": true,
    "target": "es2015",
  },
]
`;

exports[`NgJestTransformer should use esbuild to process mjs or \`node_modules\` js files to CJS codes 3`] = `
Array [
  "
      const pi = parseFloat(3.124);
      
      export { pi };
    ",
  Object {
    "format": "cjs",
    "loader": "js",
    "sourceRoot": undefined,
    "sourcefile": "foo.mjs",
    "sourcemap": true,
    "sourcesContent": true,
    "target": "es2016",
  },
]
`;

exports[`NgJestTransformer should use esbuild to process mjs or \`node_modules\` js files to CJS codes 4`] = `
Array [
  "
      const pi = parseFloat(3.124);
      
      export { pi };
    ",
  Object {
    "format": "cjs",
    "loader": "js",
    "sourceRoot": undefined,
    "sourcefile": "node_modules/foo.js",
    "sourcemap": true,
    "sourcesContent": true,
    "target": "es2016",
  },
]
`;

exports[`NgJestTransformer should use esbuild to process mjs or \`node_modules\` js files to CJS codes 5`] = `
Array [
  "
      const pi = parseFloat(3.124);
      
      export { pi };
    ",
  Object {
    "format": "cjs",
    "loader": "js",
    "sourceRoot": undefined,
    "sourcefile": "foo.mjs",
    "sourcemap": true,
    "sourcesContent": true,
    "target": "es2015",
  },
]
`;

exports[`NgJestTransformer should use esbuild to process mjs or \`node_modules\` js files to CJS codes 6`] = `
Array [
  "
      const pi = parseFloat(3.124);
      
      export { pi };
    ",
  Object {
    "format": "cjs",
    "loader": "js",
    "sourceRoot": undefined,
    "sourcefile": "node_modules/foo.js",
    "sourcemap": true,
    "sourcesContent": true,
    "target": "es2015",
  },
]
`;

exports[`NgJestTransformer should use esbuild to process mjs or \`node_modules\` js files to ESM codes 1`] = `
Array [
  "
      const pi = parseFloat(3.124);
      
      export { pi };
    ",
  Object {
    "format": "esm",
    "loader": "js",
    "sourceRoot": undefined,
    "sourcefile": "foo.mjs",
    "sourcemap": false,
    "sourcesContent": true,
    "target": "es2015",
  },
]
`;

exports[`NgJestTransformer should use esbuild to process mjs or \`node_modules\` js files to ESM codes 2`] = `
Array [
  "
      const pi = parseFloat(3.124);
      
      export { pi };
    ",
  Object {
    "format": "esm",
    "loader": "js",
    "sourceRoot": undefined,
    "sourcefile": "node_modules/foo.js",
    "sourcemap": false,
    "sourcesContent": true,
    "target": "es2015",
  },
]
`;

exports[`NgJestTransformer should use esbuild to process mjs or \`node_modules\` js files to ESM codes 3`] = `
Array [
  "
      const pi = parseFloat(3.124);
      
      export { pi };
    ",
  Object {
    "format": "esm",
    "loader": "js",
    "sourceRoot": undefined,
    "sourcefile": "foo.mjs",
    "sourcemap": true,
    "sourcesContent": true,
    "target": "es2016",
  },
]
`;

exports[`NgJestTransformer should use esbuild to process mjs or \`node_modules\` js files to ESM codes 4`] = `
Array [
  "
      const pi = parseFloat(3.124);
      
      export { pi };
    ",
  Object {
    "format": "esm",
    "loader": "js",
    "sourceRoot": undefined,
    "sourcefile": "node_modules/foo.js",
    "sourcemap": true,
    "sourcesContent": true,
    "target": "es2016",
  },
]
`;

exports[`NgJestTransformer should use esbuild to process mjs or \`node_modules\` js files to ESM codes 5`] = `
Array [
  "
      const pi = parseFloat(3.124);
      
      export { pi };
    ",
  Object {
    "format": "esm",
    "loader": "js",
    "sourceRoot": undefined,
    "sourcefile": "foo.mjs",
    "sourcemap": true,
    "sourcesContent": true,
    "target": "es2015",
  },
]
`;

exports[`NgJestTransformer should use esbuild to process mjs or \`node_modules\` js files to ESM codes 6`] = `
Array [
  "
      const pi = parseFloat(3.124);
      
      export { pi };
    ",
  Object {
    "format": "esm",
    "loader": "js",
    "sourceRoot": undefined,
    "sourcefile": "node_modules/foo.js",
    "sourcemap": true,
    "sourcesContent": true,
    "target": "es2015",
  },
]
`;
